README.md
## README - jmaintenace

### Abstract

jmaintenance is a system recovery program that tries to fix common errors which might keep the system from booting.
This program is for Debian GNU/Linux systems (tested on Wheezy 7.5, but other versions might work as well);
do not use it on other distributions!

### Project

This program is written for the JackOS Project, a Debian based, innovative GNU/Linux distribution.

### License

This is free and open source software, released under GPL version 2. See LICENSE for more information.
Copyright (C) 2014 Péter Simon.

### Current status

jmaintenance is tested a lot, but it is still not ready for production systems. Currently it's in beta stage. You can help improve it by testing jmaintenance and reporting any bugs you find.

### Documentation

The program is documented in its manual page, jmaintenance(8) (you can find it under `doc/jmaintenance.8`).

### Installation

For installation instructions see `doc/INSTALL`.

### Authors

Author: Peter Simon<br>
Contributors: Michael Caron Coutier, Ben Stevenson

### Contact

To report bugs, submit patches, ask for features or technical support, write [here](mailto:psimonprs@gmail.com).


doc/CHANGELOG
CHANGELOG - jmaintenance

NOTE: this changelog is not complete, just a history of development in a nutshell. 
      Complete revision history is available from version 0.6, where we migrated to Git.


Release 0.1-alpha	2014/01/11
	*Initial release*

Release 0.2-beta	2014/02/12
	Feature:	rootfs repair with fsck
	Feature:	/home permission and ownership repair
	Feature:	restoring $PATH to a safe value for script subshell
	Feature:	logfiles have timestamps in their names (thanks for Ben Stevenson)
	Feature:        --version option
	Feature:	wait for user confirmation on warning
	Feature:	prompt before dpkg reconfiguration
	Feature:        colorized [ INFO ] messages
	Change:		hardcoded program checks replaced with chk_opt () and chk_crit ()
	Bugfix:		typos

Release 0.3-beta	2014/03/28
	Feature:	list possible choices in repair_usperm ()
	Feature:	self-destruction protection
	Feature:	using readline input to allow backspaces
	Feature:	new options: -R -U -l <LOGFILE> -L -s -W -h
	Change:		option handling now uses getops
	Change:		performing user check BEFORE log file creation
	Change:		user check replaced with /usr/bin/whoami
	Change:		unified coding style
	Bugfix:		trying to write to readonly root
	Bugfix:		menu selects operation on special keystroke (eg. PageUp)
	Bugfix:		useless user check log entry
	Bugfix:         typos
	Other:		license changed to GPLv2
	Other:		added manpage (jmaintenance(8))

Release 0.4-beta	2014/04/17
	Feature:	ncurses main menu
	Feature:	prompt to reboot when completed
	Feature:	configuration file support
	Feature:	/usr/(s)bin permission repair
	Change:		rebuilt choice logging system
	Change:		while : changed to while true (more readable)
	Change:		reordered help message
	Change:		more sensible logic in skipping system checks
	Change:		chk_opt return value removed
	Change:		info color changed to cyan
        Change:         fixed more odd-style code
	Change:		1/0 switches replaced with "true"/"false" ones
	Change:		function declarations changed to "function samplefunc {...}"
	Bugfix:		replaced all program calls with $EXEC_PATH-style ones
	Bugfix:		warningpr() ignores $NOWARN value
	Bugfix:		exit on unknown runlevel
	Bugfix:		line 224, if statement - non-unary operator
	Bugfix:		$SCRIPTDIR determination now uses absolute paths
	Bugfix:		removed readline input from single-keystroke read calls
	Bugfix:		now using local variables
	Bugfix:		strings now double-quoted
	Other:		moved CHANGELOG and TODO to separate files
	Other:		README, LICENSE and INSTALL files created
	Other:		more detailed description in manpage
	Other:		bug "rootfs repair in recovery mode doesn't work" presented in the manpage
	Other:		cleaned up comments

Release 0.5-beta
	NOTE:		not listing all changes in detail here as 0.5 is nearly a full rewrite.
	
	Feature:	fix system directory permissions
	Feature:	reinstall all packages
	Feature:	fix packages with apt-get -f install
	Feature:	now more messages are in ncurses
	Feature:	more sophisticated fs_repair function
	Bugfix:		unified CLI options
	Change:		merged chk_crit and chk_opt functions into chk_exec PRIORITY TARGET
	Change:         single log file
	Change:         more readable [ OK|INFO|WARN|FAIL ] messages
	Change:		removed feature: self-removal protection (reason: useless, the script resides under /usr/sbin/jmaintenace)
	Other:		code split into multiple files
	Other:		tarball directory tree changed
	Other:		full code cleanup

Release 0.6-beta
	Feature:	installer script
	Other:		complete commit history throught Git, this file is no longer maintained

src/main.bash
#!/bin/bash
####################
### jmaintenance ###
####################
# main.bash - main source file
#
# Copyright (C) 2014 Peter Simon
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#


### BEGIN SCRIPT ###
SCRIPTPATH="/opt/jmaintenance/"
if cd $SCRIPTPATH 2> /dev/null;
then
    :
else
    printf "\e[7;31mFAIL\e[0m Failed to change working directory!\nMake sure you are root and the program resides in $SCRIPTPATH.\nAbort.\n"
    exit 71 # OSERR
fi

# Function to include files # USAGE: include FILE
function include
{
    local target="$1"
    if source "$target";
    then
	:
    else
	printf "ERROR! failed to include $target, abort!\n"
	exit 71 # OSERR
    fi
}

# Executable paths
include "exec_paths.bash"

# Configuration and option handling
include "config.bash"
while getopts "Ll:URWVhv" "option"
do
    case "$option" in
	"L")
	    LOGGING="false"
	    ;;
	"l")
	    LOGGING="true"
	    LOGFILE="$OPTARG"
	    ;;
	"U")
	    SKIP_USRCHK="true"
	    ;;
	"R")
	    SKIP_RLCHK="true"
	    ;;
	"W")
	    NOWARN="true"
	    ;;
	"V")
            VERBOSE="false"
	    exec 2>/dev/null
            ;;
	"h")
	    printf "Usage: maintenance [OPTIONS]\n"
            printf "Perform system maintenance and recovery in single-user mode.\n\n"
            printf "Options:\n"
            printf " -l <LOGFILE>\tspecify logfile location\n"
            printf " -L\t\tdisable logging\n"
            printf " -R\t\tskip runlevel check (USE WITH CAUTION!)\n"
            printf " -U\t\tskip user check\n"
            printf " -W\t\tdisable warning messages (USE WITH CAUTION!)\n"
            printf " -V\t\tdisble verbose output (silent)\n"
            printf " -h\t\tdisplay this help and exit\n"
            printf " -v\t\tdisplay version information and exit\n"
            exit $EX_OK
	    ;;
	"v")
	    printf "jmaintenance $VERSION\n"
            printf "Copyright (C) 2014 Péter Simon\n"
            printf "License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>.\n"
            printf "This is free software: you are free to change and redistribute it.\n"
            printf "There is NO WARRANTY, to the extent permitted by law.\n\n"
            printf "Written by Péter Simon.\n"
            exit $EX_OK
            ;;
	*)
            exit $EX_USAGE
            ;;
    esac
done

# Basic function definitions
include "basic_functions.bash"

# Create log file
if [ "$LOGGING" == "true" ]
then
    : >> "$LOGFILE"
    if [ $? -eq 0 ]
    then
	LOGGING="true"
	message "optional" "scroll" "OK" "Logfile created"
	date="$($DATE_PATH +"%Y %m/%d %H:%M")"
	log "\n$INFO jmaintenance started at $date"
    else
	LOGGING="false"
	message "critical" "warning" "WARN" "Failed to create logfile, no logging!"
    fi
else
    message "optional" "scroll" "INFO" "Not creating logfile"
fi

## System checks ##

# Pass 1 - check chmod
if [ -e "$CHMOD_PATH" ]
then
    log "$OK $CHMOD_PATH exists"
    message "optional" "scroll" "OK" "$CHMOD_PATH exists"
else
    log "$FAIL $CHMOD_PARH doesn't exist, abort!"
    message "critical" "halt" "FAIL" "$CHMOD_PATH doesn't exist, abort!"
    exit $EX_OSFILE
fi
if [ -x "$CHMOD_PATH" ]
then
    log "$OK $CHMOD_PATH is executable"
    message "optional" "scroll" "OK" "$CHMOD_PATH is executable"
else
    log "$WARN $CHMOD_PATH is not executable, attempting setfacl!"
    message "critical" "scroll" "WARN" "$CHMOD_PATH is not executable, attempting setfacl!"
    $SETFACL_PATH --set u::rwx,g::r-x,o::r-x "$CHMOD_PATH" 2>/dev/null
    SETFACL_EXIT=$?
    if [ $SETFACL_EXIT -eq 0 ]
    then
	log "$OK setfacl succeeded"
	message "critical" "scroll" "OK" "setfacl succeeded"
    else
	log "$WARN setfacl aborted with error code $SETFACL_EXIT, attempting to install acl"
        message "critical" "scroll" "WARN" "setfacl aborted, attempting to install package acl..."
	$APT_GET_PATH update > /dev/null
	$APT_GET_PATH install acl > /dev/null
	if [ $? -eq 0 ]
	then
	    log "$OK installed acl, rerunning setfacl"
	    message "optional" "scroll" "OK" "installed acl, rerunning setfacl"
	    $SETFACL_PATH --set u::rwx,g::r-x,o::r-x "$CHMOD_PATH" 2>/dev/null
	    SETFACL_EXIT=$?
	    if [ $SETFACL_EXIT -eq 0 ]
	    then
		log "$OK setfacl succeeded"
		message "critical" "scroll" "OK" "setfacl succeeded"
	    else
		log "$FAIL setfacl failed, abort!"
		message "critical" "halt" "FAIL" "setfacl failed, abort!"
		exit $EX_OSERR
	    fi
	else
	    log "$FAIL failed to install acl, abort!"
	    message "crtical" "halt" "FAIL" "Failed to install acl, abort!"
	    exit $EX_OSERR
	fi
    fi
fi
# Pass 2 - Check whoami and user
if [ "$SKIP_USRCHK" == "false" ]
then
    chk_exec "optional" "$WHOAMI_PATH"
    if [ $? -eq 0 ]
    then
	if [ "$($WHOAMI_PATH)" == "root" ]
	then
	    message "optional" "scroll" "OK" "Running as root"
	    log "$OK Running as root"
	else
	    message "critical" "scroll" "FAIL" "Not running as root, abort!"
	    log "$FAIL Not running as root, abort!"
	    exit $EX_NOPERM
	fi
    else
	log "$WARN User check skipped!"
	message "critical" "scroll" "WARN" "User check skipped!"
    fi
else
    log "$WARN User check skipped!"
    message "critical" "scroll" "WARN" "User check skipped!"
fi
# Pass 3 - Check runlevel
if [ "$SKIP_RLCHK" == "false" ]
then
    chk_exec "optional" "$RUNLEVEL_PATH"
    chk_exec "optional" "$AWK_PATH"
    if [ "$($RUNLEVEL_PATH | $AWK_PATH '{print $2}')" == "S" ]
    then
	log "$OK System is in single-user mode"
	message "optional" "scroll" "OK" "System is in single-user mode"
    else
	log "$FAIL System is not in single-user mode, abort!"
	message "critical" "halt" "FAIL" "System is not in single-user mode, abort!"
	exit $EX_OSERR
    fi
else
    log "$WARN Runlevel check skipped!"
    message "critical" "scroll" "WARN" "Runlevel check skipped!"
fi

# Repairing functions
include repair_functions.bash

# Start menu
include menu.bash

src/menu.bash
#!/bin/bash
####################
### jmaintenance ###
####################
# menu.bash - main menu
#
# Copyright (C) 2014 Peter Simon
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#


### BEGIN SCRIPT ###

function input
{
    printf "\n"
    message "critical" "scroll" "INFO" "Press RETURN to go back to menu"
    read -n1 -s
}

while true; do
    if [ "$FRONTEND" == "ncurses" ]
    then
	$WHIPTAIL_PATH --backtitle "jmaintenance $VERSION" --title "Main Menu" \
	    --cancel-button "Quit" \
	    --menu "Select an action (it is recommended to do them in order)" \
	    0 0 0\
	    "a" "Autopilot (do all tasks)" \
	    "1" "Fix filesystems" \
	    "2" "Fix broken packages" \
	    "3" "Fix system permissions" \
	    "4" "Reinstall all packages" \
	    "5" "Restore system configuration" \
	    "6" "Fix user permissions" \
	    "7" "Free up disk space" \
	    "q" "quit" \
	    2>/tmp/mmch
	choice="$($CAT_PATH /tmp/mmch)"
	printf "\n\n"
    else
	$CLEAR_PATH 2> /dev/null
	printf \
"###############################\n\
#### jmaintenance $VERSION ####\n\
###############################\n\
a Autopilot (do all tasks)\n\
1 Fix filesystems\n\
2 Fix broken packages\n\
3 Fix system permissions\n\
4 Reinstall all packages\n\
5 Restore system configuration\n\
6 Fix user permissions\n\
7 Free up disk space\n\
q quit\n\n\
Choice: "
	local choice
	read -n1 -s choice
	printf "\n"
    fi
    
    case "$choice" in
	"a")
	    fs_repair
	    brokenpkg_repair
	    fix_sysperm
	    sys_reinstall
	    reconfigure
	    fix_usperm
	    freeupdisk
	    ;;
	"1")
	    fs_repair
	    ;;
	"2")
	    brokenpkg_repair
	    ;;
	"3")
	    fix_sysperm
	    ;;
	"4")
	    sys_reinstall
	    ;;
	"5")
	    reconfigure
	    ;;
	"6")
	    fix_usperm
	    ;;
	"7")
	    freeupdisk
	    ;;
	"q")
	    $RM_PATH /tmp/mmch
	    exit $EX_OK
	    ;;
	"")
	    $RM_PATH /tmp/mmch
	    exit $EX_OK
	    ;;
    esac
    input
done

src/config.bash
#!/bin/bash
####################
### jmaintenance ###
####################
# config.bash - configuration
#
# Copyright (C) 2014 Peter Simon
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

### BEGIN SCRIPT ###

## Pass 1 - default options (readonly variables cannot be overridden by CLI of config file options) ##

# Version info
readonly VERSION="0.6.1-beta"

# $PATH
readonly PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"

# Configuration file location
readonly CONFIGFILE="/etc/jmaintenance.conf"

# Log file location
LOGFILE="/var/log/jmaintenance"

# Logging
LOGGING="true"

# Frontend
if [ -x "$WHIPTAIL_PATH" ]
then
    FRONTEND="ncurses"
else
    $CHMOD_PATH 755 $WHIPTAIL_PATH 2> /dev/null
    if [ $? -eq 0 ]
    then
	FRONTEND="ncurses"
    else
	FRONTEND="text"
    fi
fi
readonly FRONTEND

# Verbosity
VERBOSE="true"

# Display warning messages
NOWARN="false"

# Skip user and runlevel checks?
SKIP_USRCHK="false"
SKIP_RLCHK="false"

# Exit codes (from /usr/include/sysexits.h)
EX_OK=0
EX_USAGE=64
EX_OSERR=71
EX_OSFILE=72
EX_NOPERM=77

# Sample messages
readonly OK="\e[7;32m OK \e[0m"
readonly INFO="\e[7;36mINFO\e[0m"
readonly WARN="\e[7;33mWARN\e[0m"
readonly FAIL="\e[7;31mFAIL\e[0m"

## Pass 2 - include options from config file ##

if [ -f "$CONFIGFILE" ]
then
    source "$CONFIGFILE"
fi

## Pass 3 - option handling happens in main source file ##

src/exec_paths.bash
#!/bin/bash
####################
### jmaintenance ###
####################
# exec_paths.bash - paths of the executables used
#
# Copyright (C) 2014 Peter Simon
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

### BEGIN SCRIPT ###

# date
DATE_PATH="/bin/date"
# whiptail
WHIPTAIL_PATH="/usr/bin/whiptail"
# chmod
CHMOD_PATH="/bin/chmod"
# chown
CHOWN_PATH="/bin/chown"
# dirname
DIRNAME_PATH="/usr/bin/dirname"
# pwd
PWD_PATH="/bin/pwd"
# whoami
WHOAMI_PATH="/usr/bin/whoami"
# runlevel
RUNLEVEL_PATH="/sbin/runlevel"
# awk
AWK_PATH="/usr/bin/awk"
# fsck
FSCK_PATH="/sbin/fsck"
# find
FIND_PATH="/usr/bin/find"
# rm
RM_PATH="/bin/rm"
# dpkg
DPKG_PATH="/usr/bin/dpkg"
# dpkg-reconfigure
DPKG_RECONFIGURE_PATH="/usr/sbin/dpkg-reconfigure"
# setfacl
SETFACL_PATH="/bin/setfacl"
# wc
WC_PATH="/usr/bin/wc"
# grep
GREP_PATH="/bin/grep"
# cat
CAT_PATH="/bin/cat"
# apt-get
APT_GET_PATH="/usr/bin/apt-get"
# mount
MOUNT_PATH="/bin/mount"
# shutdown
SHUTDOWN_PATH="/sbin/shutdown"
# xargs
XARGS_PATH="/usr/bin/xargs"
# aptitude
APTITUDE_PATH="/usr/bin/aptitude"
# clear
CLEAR_PATH="/usr/bin/clear"
# ping
PING_PATH="/bin/ping"
# ifup
IFUP_PATH="/sbin/ifup"

src/repair_functions.bash
#!/bin/bash
####################
### jmaintenance ###
####################
# repair_functions.bash - definition of the fixing functions
#
# Copyright (C) 2014 Peter Simon
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

### BEGIN SCRIPT ###


# Repair filesystems with fsck
function fs_repair
{
    log "$INFO fs_repair called"
    
    # Check executables
    local target
    for target in "$FSCK_PATH" "$CAT_PATH" "$GREP_PATH" "$AWK_PATH" "$MOUNT_PATH"; do
	! chk_exec "critical" "$target"  && return 1
    done

    # Function to check fsck exit code
    function chk_exit
    {
	case $FSCK_EXIT in
	    0)
		message "critical" "scroll" "OK" "Filesystem $device has no errors"
		;;
	    1)
		message "critical" "scroll" "OK" "Errors on filesystem $device corrected"
		;;
	    4)
		message "critical" "warning" "FAIL" "Unable to correct filesystem errors on $device!"
		;;
	    3|6)
                case $FSCK_EXIT in
		    3)
			message "critical" "halt" "OK" "Errors on filesystem $device corrected, rebooting system!\nAfter reboot, rerun jmaintenance if you haven't completed all tasks\n(this should be the case if you've chosen autopilot)."
			;;
		    6)
			message "critical" "halt" "FAIL" "Failed to correct errors on filesystem $device, rebooting system!\nAfter reboot, rerun jmaintenance if you haven't completed all tasks\n(this should be the case if you've chosen autopilot).\nIf this step fails on the next run as well, you may have a hardware error."
			;;
		esac
		$MOUNT_PATH -o remount,rw "$device"
		if [ $? -eq 0 ]
		then
		    message "optional" "scroll" "OK" "Filesystem $device remounted read-write"
		    log "$OK fsck on $device finished with exit code $FSCK_EXIT, rebooting system..."
		else
		    message "critical" "scroll" "FAIL" "Failed to remount $device read-write!"
		fi
		message "optional" "scroll" "INFO" "Rebooting system..."
		chk_exec "critical" "$SHUTDOWN_PATH"
		if [ $? -eq 0 ]
		then
		    $SHUTDOWN_PATH -r now
		else
		    message "critical" "halt" "WARN" "$SHUTDOWN_PATH not found, please manually reboot the system!"
		    exit $EX_OSERR
		fi
		;;
        esac
    }

    if [ -e /etc/fstab ]
    then
	log "$OK fstab found, remounting filesystems readonly"
	log "$INFO disabling logging temporarily"
	local device
	for device in $($CAT_PATH /etc/fstab | $GREP_PATH "/" | $GREP_PATH --invert-match "#" | $GREP_PATH --invert-match "/media" | $AWK_PATH '{print $2}'); do
	    $MOUNT_PATH -o remount,ro "$device"
	    if [ $? -eq 0 ]
	    then
		message "optional" "scroll" "OK" "Remounted $device readonly"
		$FSCK_PATH -y -f -T -C "$device"
		local FSCK_EXIT=$?
		chk_exit
		$MOUNT_PATH -o remount,rw "$device"
                if [ $? -eq 0 ]
                then
                    message "optional" "scroll" "OK" "Filesystem $device remounted read-write"
		    log "$OK fsck on $device finsihed with exit code $FSCK_EXIT, remounted filesystem"
		else
		    message "critical" "warning" "FAIL" "Failed to remount $device read-write!"
		fi
	    else
		message "critical" "scroll" "FAIL" "Failed to remount $device, not repairing it!"
	    fi
	done
	return 0
    else
	log "$FAIL fstab not found, abort!"
	message "critical" "halt" "FAIL" "fstab not found, abort!"
	return 1
    fi
}

# Repair broken pacakges with dpkg
function brokenpkg_repair
{
    log "$INFO brokenpkg_repair called"

    # check dpkg
    ! chk_exec "critical" "$DPKG_PATH" && return 1
    
    # Pass 1 - dpkg
    $DPKG_PATH --configure -a
    local DPKG_EXIT=$?
    if [ $DPKG_EXIT -eq 0 ]
    then
	log "$OK dpkg --configure -a finished"
	message "critical" "scroll" "OK" "dpkg configuration finished"
    else
	log "$FAIL dpkg --configure -a failed with exit code $DPKG_EXIT"
	message "critical" "scroll" "FAIL" "dpkg failed!"
    fi
    
    # check apt-get
    ! chk_exec "critical" "$APT_GET_PATH" && return 1

    # Pass 2 - apt-get
    $APT_GET_PATH -f install
    local APT_EXIT=$?
    if [ $DPKG_EXIT -eq 0 ]
    then
        log "$OK apt-get -f install finished"
        message "critical" "scroll" "OK" "apt fix finished"
    else
        log "$FAIL apt-get -f install failed with exit code $DPKG_EXIT"
        message "critical" "scroll" "FAIL" "apt fix failed!"
    fi
}

# Restore system directory permissions
function fix_sysperm
{
    log "$INFO fix_sysperm called"
    ! chk_exec "critical" "$CHMOD_PATH" && return 1

    local target
    for target in "/bin" "/sbin" "/usr" "/usr/bin" "/usr/bin" "/boot" "/dev" "/etc" "/home" "/lib" "/media" "/mnt" "/opt" "/run" "/srv" "/sys" "/var"
    do
       	if $CHMOD_PATH 755 "$target";
	then
	    log "$OK restored $target permissions"
	    message "critical" "scroll" "OK" "Restored $target permissions"
	else
	    log "$FAIL failed to restore /bin permissions!"
	    message "critical" "warning" "FAIL" "Failed to restore $target permissions!"
	fi
    done
}

# Reinstall all packages
function sys_reinstall
{
    log "$INFO sys_reinstall called"

    # Bring up networking
    if /etc/init.d/networking start;
    then
	log "$OK Networking started"
	message "optional" "scroll" "OK" "Networking started"
    else
	log "$FAIL Failed to start networking, abort!"
	message "critical" "scroll" "FAIL" "Failed to start networking, abort!"
	return
    fi
    for iface in "$($CAT_PATH /etc/network/interfaces | $GREP_PATH --invert-match lo| $GREP_PATH iface | $AWK_PATH '{print $2}')"
    do
	if $IFUP_PATH "$iface";
	then
	    log "$OK Brought up interface $iface, abort!"
	    message "optional" "scroll" "OK" "Brought up interface $iface"
	else
	    log "$FAIL Failed to bring up interface $iface, abort!"
	    message "critical" "scroll" "FAIL" "Failed to bring up interface $iface, abort!"
	fi
    done
    
    log "$INFO Testing network connection..."
    message "optional" "scroll" "INFO" "Testing network connection..."
    if $PING_PATH -c 1 debian.org > /dev/null;
    then
	log "$OK Networking is working"
	message "optional" "scroll" "OK" "Networking is working"
    else
	log "$FAIL Networking is not working!"
	message "critical" "warning" "FAIL" "Networking is not working, unable to reinstall system!"
    fi
    ! chk_exec "critical" "$AWK_PATH" && return 1
    ! chk_exec "critical" "$DPKG_PATH" && return 1
    ! chk_exec "critical" "$XARGS_PATH" && return 1
    ! chk_exec "critical" "$APTITUDE_PATH" && return 1
    
    message "critical" "halt" "INFO" "Now we are going to reinstall all of the packages in order to restore the default permissions. It might take some hours if you have large packages like GNOME or KDE installed.";
    
    $DPKG_PATH --get-selections \* | $AWK_PATH '{print $1}' | $XARGS_PATH -l1 $APTITUDE_PATH reinstall
    local APTITUDE_EXIT=$?
    if [ $APTITUDE_EXIT -eq 0 ]
    then
	log "$OK Reinstallation succeeded"
	message "optional" "scroll" "OK" "Reinstallation succeeded"
    else
	log "$FAIL Reinstallation failed!"
	message "critical" "warning" "FAIL" "Reinstallation failed!"
    fi
}

# Reconfigure all packages with
function reconfigure
{
    log "$INFO reconfigure called"

    # check dpkg-reconfigure
    ! chk_exec "critical" "$DPKG_RECONFIGURE_PATH" && return 1
    
    message "critical" "halt" "INFO" "Now we are going to reconfigure all of your installed packages. Some questions may be asked during this. If you don't know what to answer, the default choice is usually a good one."
    $DPKG_RECONFIGURE_PATH -pcritical --all --force
    local DPKG_RECONFIGURE_EXIT=$?
    printf "\n"
    if [ $DPKG_RECONFIGURE_EXIT -eq 0 ]
    then
	log "$OK dpkg-reconfigure -pcritical --all -force finished"
	message "critical" "scroll" "OK" "reconfiguration finished"
    else
	log "$FAIL dpkg-reconfigure -pcritical --all -force failed with error code $DPKG_RECONFIGURE_EXIT"
	message "critical" "scroll" "FAIL" "reconfiguration failed!"
    fi
}

# Restore user permissions
function fix_usperm
{
    log "$INFO fix_usperm called"

    ! chk_exec "critical" "$FIND_PATH" && return 1
    ! chk_exec "critical" "$CHMOD_PATH" && return 1
    ! chk_exec "critical" "$CHOWN_PATH" && return 1
    ! chk_exec "critical" "$XARGS_PATH" && return 1
    ! chk_exec "critical" "$GREP_PATH" && return 1
    ! chk_exec "critical" "$AWK_PATH" && return 1

    message "critical" "halt" "INFO" "Now we're going to restore the permissions of your personal files\n(the ones under /home)."

    # Pass 1 - file modes
    if $FIND_PATH /home/ -type f -print0 | $XARGS_PATH -0 $CHMOD_PATH 0644;
    then
	log "$OK restored permissions of files under /home"
	message "critical" "scroll" "OK" "Restored permissions of files under /home"
    else
	log "$FAIL failed to restore permissions of files under /home"
	message "critical" "scroll" "FAIL" "Failed to restore permissions of files under /home!"
    fi

    # Pass 2 - directory modes 
    if $FIND_PATH /home/ -type d -print0 | $XARGS_PATH -0 $CHMOD_PATH 0755;
    then
        log "$OK restored permissions of directories under /home"
	message "critical" "scroll" "OK" "Restored permissions of directories under /home"
    else
        log "$FAIL failed to restore permissions of directories under /home"
	message "critical" "scroll" "FAIL" "Failed to restore permissions of directories under /home!"
    fi

    # Pass 3 - ownerships
    local target
    for target in $($CAT_PATH /etc/passwd | $GREP_PATH /home/ | $AWK_PATH -F : '{print $1}')
    do
	if $CHOWN_PATH -R "$target":"$target" /home/"$target"/;
	then
	    log "$OK Fixed ownerships of user $target"
	    message "critical" "scroll" "OK" "Fixed ownerships of user $target"
	else
	    log "$FAIL Failed to fix ownerships of user $target!"
	    message "critical" "scroll" "FAIL" "Failed to fix ownerships of user $target!"
	fi
    done    
}

# Free up disk space
function freeupdisk
{
    log "$INFO freeupdisk called"

    ! chk_exec "critical" "$APT_GET_PATH" && return 1
    ! chk_exec "critical" "$RM_PATH" && return 1

    # Pass 1 - clear APT cache
    if $APT_GET_PATH clean;
    then
	log "$OK Cleaned APT cache"
	message "critical" "scroll" "OK" "Cleaned APT cache"
    else
	log "$FAIL Failed to clean APT cache!"
	message "critical" "scroll" "FAIL" "Failed to clean APT cache!"
    fi

    # Pass 2 - KDE, desktop, thumbnails cache
    local target
    for target in "/home/*/.kde/cache-*" "/home/*/.cache" "/home/*/.thumbs" "/home/*/.thumbnails"
    do
	if [ -d $target ]
	then
	    if $RM_PATH -rf $target;
	    then
		log "$OK Cleared $target"
		message "critical" "scroll" "OK" "Cleared $target"
	    else
		log "$FAIL Failed to clear $target!"
		message "critical" "scroll" "FAIL" "Failed to clear $target!"
	    fi
	fi
    done
}